### DISCUSSION TITLE
Incomplete dates

### DISCUSSION THREAD

**Discussion by chrishas35** [immich-app/immich#6071] [General]
For some photos, pre-digital or digital where original metadata has been lost, I may not have the precise date/time it was taken.  I'd still like these to have some level of date information though. ISO 8601 supports various level of partial dates (YYYY or YYYY-MM), but it'd also be nice to say "it was in the YYX0 decade" or even "it was in the YY00 century", so associating some level of precision to the fuzzy date would be good. One step further though I could also see benefit in providing a "fuzzy" range, for example you know they lived in this house from YYYY-ZZZZ but not much more beyond that can be determined from the photo.

It would also be beneficial able to sort photos relative to one another when there are incomplete dates.  This could either be against other incomplete date photos, or (and this may be aided by range support) against a photo with a complete date.

**Comment by aisbergde** [comment]
I don't get the whole idea.
I also have this kind of assets, but what would be the features in immich?
- Search and find them?
- How to sort them in the timline, if they should represent a range?
- about comparing with other assets at the same time: I would say, this could be a job for some external software to manage photos and their metadata. How this should or could look in immich? immich is showing thumbnails or single photos.

**Reply by chrishas35** [reply]
My main point would be the remove ambiguity from the date.  The solution I've seen most suggest for this is to just decide to pick a date, for example, January 1, YYYY to represent a year. But immich will display that full date (and time for that matter) and it becomes a "was this really January 1, YYYY or was it some time in YYYY?" IMO this really starts to break down with wider ranges, such as multiple years. If there are other workaround, I'm happy to give them a shot for the use-case, but the one I see over and over is less than ideal IMO.

Even an external solution managing the metadata can't really solve that ambiguous display issue as immich is always going to display out the full date/time currently. This may be viable if, at a minimum, it was able to provide a precision to a date (day, month, year, decade, etc.) and adjusting the immich display accordingly would be helpful. However, as noted this breaksdown for something that spans multiples.

**Comment by dkbarn** [comment]
I would also like to see this feature, as I use Immich to store digitized scans of old photos. For most of these photos, the exact date is unknown, and can only be inferred from context or maybe from a caption, but these captions often say things like:

* Christmas 1967
* Spring 1971
* January 1980

I want to capture these dates in Immich, so that I can sort and search for them by date, but I don't want to arbitrarily make-up fake values for timestamp components that I don't actually know, simply because Immich requires that a complete YYYY-MM-DD HH:MM:SS be provided.

The best way I can think to support this is to allow for an upper and lower bound of each photo timestamp, to define a possible range of dates.

In terms of how searching and sorting would function with this feature, it seems to me that searching would be pretty straightforward -- you can still search for a date and if that date appears somewhere within the min - max range of a given photo then it matches the search.
For sorting, you could choose to sort photos by their min, max, or midpoint value. I don't think it's especially important which. For photos that have a known timestamp, the min and max are the same, so it makes no difference.

**Comment by radh21301** [comment]
I find this feature request very confusing. And in my opinion, it's not something that an app like Immich is supposed to solve.

Firstly, Immich is not really enforcing the YYYY-MM-DD -HH-mm-ss rule, is it? The EXIF field requires that structure (you could force the field to take partial dates, but many programs don't accept partial dates), and immich has to respect that. Say Immich goes ahead and implements what you ask (Summer of 1965 or something), then it cannot be stored in either the asset's EXIF field or the accompanying XMP sidecar file. What this would mean is that -> whatever edits you make with Immich has to be stored in the Immich database, and cannot be used with other tools if you decide to migrate. That is not a good design.

To me, the best solution looks like setting the date like you suggested -> The best guess you can make. If you have no estimate, then January 01, YEAR, midnight (00:00:00). If you have a series of photos you digiitzed, and want to keep them in the correct order-> add a second to each photo (you can ask exiftool to sort by filename and add one second to each image). This would keep all your images in the correct order when you view them on the Immich timeline. 

And if you want to say that it was Spring of 1995 or something, just add it to the description. OR add a tag that says 'date is tentative' or something. Description and Tags can be stored in the asset's metadata, so it will be available when you migrate.


References:
https://exiftool.org/forum/index.php?topic=11575.0
https://exiftool.org/forum/index.php?topic=3956.0
https://news.ycombinator.com/item?id=34377584

**Reply by dkbarn** [reply]
> in my opinion, it's not something that an app like Immich is supposed to solve.
> Immich is not really enforcing the YYYY-MM-DD -HH-mm-ss rule, is it? The EXIF field requires that structure

It is precisely because EXIF does not support this concept natively that I believe Immich is where the solution should live. Immich has a database, it has the ability to capture more sophisticated metadata than EXIF can. Why not leverage that fact to support features that aren't possible with standard EXIF fields alone?

> Say Immich goes ahead and implements what you ask (Summer of 1965 or something), then it cannot be stored in either the asset's EXIF field or the accompanying XMP sidecar file. What this would mean is that -> whatever edits you make with Immich has to be stored in the Immich database, and cannot be used with other tools if you decide to migrate. That is not a good design.

Ok, suppose a different approach is taken. Rather than storing a timestamp range, which cannot be translated into anything EXIF can understand, you instead store a timestamp + accuracy value. For example:

* 1967-12-01 00:00:00 + accuracy "MONTH"
* 1995-03-17 00:00:00 + accuracy "DAY"
* 1980-01-01 00:00:00 + accuracy "YEAR"

The accuracy is used by Immich to mask out the values of the timestamp beyond that component. Now you have a full valid timestamp which is compatible with EXIF and the XMP sidecar files, and can be stored there for use by other tools, and you have an extra piece of metadata (the accuracy value) which only Immich understands, and can be stored in its database or as a custom EXIF field for its own use.

**Reply by radh21301** [reply]
I see what you are saying, but I am not totally sure I understand the use case. 

Say Immich adds that accuracy field like you mentioned, where would that be used? The way Immich orders the images would still be based in the EXIF data correct?

**Reply by radh21301** [reply]
Or rather - a better question would be. What problem does this feature solve? What advantages do you get? 
Are you only looking for Immich to tell you that these dates are tentative?

**Reply by dkbarn** [reply]
The goal here is to have a way of keeping track of incomplete dates, and visually distinguishing them from dates that are complete. I agree that deciding how complete and incomplete dates should be compared when sorting is a challenge, but the exact mechanics of this I think are less important so long as a date of "1980-01-XX" appears somewhere in the timeline alongside other photos in that month.

I agree with you that I could solve this problem by simply inventing my own convention with what's already available -- for example, represent January 1980 as 1980-01-01 00:00:00 with the text "January 1980" written into the description or in a tag, or by incrementing the seconds to :01 to signal to myself that this is an estimated date. But these are all hacks and workarounds, so the intent behind this discussion is to request that Immich support this concept officially, to eliminate the need for hacks or workarounds.

**Reply by radh21301** [reply]
Okay. What do you mean by visually distinguishing them - a small icon or something on the image thumbnail? something like that?  Wondering how that distinguishing factor helps you?  Sorry, I am not able to clearly envision how this feature would look like inside Immich. Maybe someone with more experience can add to this.

P.S. From my experience, I have found that it is essential for photo management apps to stick to established standards. Purely for selfish reasons, the app you use today might not be there 10 years from now. So, any edit you make that is not a standard would essentially be lost. So, bending our individual workflows (those hacks) to accommodate these standards makes it future-proof.

Edit: I actually see your use case when I think of an image that was taken in the 1800s. But now it could be 1899 or 1800..That is a huge spread! But still I am not able to think of a future proof way - other than setting the date to 1800 Jan 01 midnight AND adding some information to the description

**Reply by aisbergde** [reply]
what I could imagine:
if the time is 00:00:00 there could be an optical indicator, that the asset has only a date but not a time.
and if it is the 01 day of the month without a time, they user could interpreted this as month without date. But this can't be a general concept: how to distinguish between day accuracy, month accuracy or year accuracy?

Maybe the combination with an accuracy tag could be used. But I agree, it would create a dependency from immich.

A more flexible way would be to define a list of tags (and optionally associated icons) to be shown on assets. This would allow an optical indicator - for whatever user defined tags.

**Reply by dkbarn** [reply]
> What do you mean by visually distinguishing them - a small icon or something on the image thumbnail? something like that?

Actually I just meant that the way that Immich displays timestamps should reflect the accuracy value that has been set for it.

Instead of this:
![image](https://github.com/user-attachments/assets/190fa4c4-3f6c-4d91-897c-2a22e164ccf1)

It would show as this:
![image](https://github.com/user-attachments/assets/a3663fef-c5d2-4f19-a60a-675a2975a2cd)

so that viewers can see that this photo in fact was not taken at midnight on New Year's Day, it was simply taken sometime in the month of January.

**Reply by CorentinB** [reply]
> It would show as this: ![image](https://private-user-images.githubusercontent.com/23280179/363307869-a3663fef-c5d2-4f19-a60a-675a2975a2cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwNTEzMTksIm5iZiI6MTczNTA1MTAxOSwicGF0aCI6Ii8yMzI4MDE3OS8zNjMzMDc4NjktYTM2NjNmZWYtYzVkMi00ZjE5LWE2MGEtNjc1YTI5NzVhMmNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDE0MzY1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5YjNmMjY0OTQwNWI3NjgyMWI5Yjk5MDVhZTZmMjdlZmQwMGI2MjhmYjg5ODhlNmI3YjI3NTVjYzZiODg1M2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P4VuDLM0qeO8v1VIWptPBo6FrvC1yasGlOZ-dWy8Oag)
> 
> so that viewers can see that this photo in fact was not taken at midnight on New Year's Day, it was simply taken sometime in the month of January.

This would be absolutely great, good example. I have many such cases where this would be very useful, because I use Immich to store and sort family pictures that I digitize.

**Reply by robertoinzerillo** [reply]
> Actually I just meant that the way that Immich displays timestamps should reflect the accuracy value that has been set for it.
> 
> It would show as this: ![image](https://private-user-images.githubusercontent.com/23280179/363307869-a3663fef-c5d2-4f19-a60a-675a2975a2cd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgwMTQ2NDcsIm5iZiI6MTc1ODAxNDM0NywicGF0aCI6Ii8yMzI4MDE3OS8zNjMzMDc4NjktYTM2NjNmZWYtYzVkMi00ZjE5LWE2MGEtNjc1YTI5NzVhMmNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE2VDA5MTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwNGM0ZjE2ZGFiOTNlYmJjNjA1MjgzNzMzYWQ1MWRmYzU0OThlNjQyMzMwZjk3YThhMTk0NmMyMWMyMTNkYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1T-0VPzXRVwJUbHkpIFu8A6ulMXWfZs0k2GNdWvlo7g)
> 
> so that viewers can see that this photo in fact was not taken at midnight on New Year's Day, it was simply taken sometime in the month of January.

I'd really like this as a standard feature.

It would solve a huge amount of problems regarding the timeline view of pictures with incomplete Date Time exif tags, specially for all those photos that don't really have any consistent/complete info (old scanned pictures, images received via whatsapp ...).

**Comment by Silver0006** [comment]
Immich uses ISO 8601 for date & time. It looks like the best option would be adding flags/bools to represent unknowns or inaccuracies

As far as I can tell the files that would need changed are:
* server/src/utils/date.spec.ts
* server/src/utils/date.ts

Then anything that depends on these files would need double checked. I am unfamiliar with immich's code if someone could point out anything I missed that would be great. 

Another option which was mentioned above was using metadata tags. Does a standard exist for this that would be accepted by other programs?

**Reply by lekorrigan** [reply]
> Does a standard exist for this that would be accepted by other programs?

Not that I know. From discussions on the Mylio forum (a rare software that allows such fuzzy dates), there is no standard, so each software has to go with their own way of doing it.

**Reply by Silver0006** [reply]
Currently working on ideas
![image](https://github.com/user-attachments/assets/9c1fe877-ff6d-43e5-8f2a-c10f11dbba23)

**Reply by lekorrigan** [reply]
Very cool.
From a user's perspective, I like when it matches workflows that deal with uncertain dates, especially with old family photos, with common cases being: 
- scribbled dates at the back, often just a month or a year ("August 1982")
- date ranges (grandma telling me this photo was when she was in Second grade at school, so it was between September 1935 and June 1936)
- identifiable season and year (there's snow so it was winter, back says a year). Mylio allows users to set custom season start and end dates for this.
- Just a decade ("That's clearly when great-grandpa was in his 50s!").
- And also "no date", when there is really nothing to help. 

All of this has proven super useful with genealogy, and for taking quick notes when batch scanning old photos and then when talking with family members with long memories. 
Mylio gets this right, but it all falls apart when exporting. One workaround when exporting is that I have to manually set the date to the beginning of the time period (so midnight on the 1 January 1982 means "1982").

I wonder if genealogy programmes like Gramps can be useful here, see https://gramps-project.org/wiki/index.php/Gramps_6.0_Wiki_Manual_-_Entering_and_editing_data:_detailed_-_part_1#Editing_dates

**Reply by jrasm91** [reply]
I would say that we need to keep the date time as is in Immich. But, we could add some _supplemental_ data that lets the user specify _additionally_  information, like a level of fuziness or confidence related to the date. I think it would be OK if the Immich clients used that information to alter how the date time was displayed. The big question seems to be whether that fuziness/confidence level should be stored as an enum "day", "month", "year", etc. or a set of start/end dates.

**Reply by Silver0006** [reply]
> I would say that we need to keep the date time as is in Immich.

Definitely agree. I think the best method going forward is adding flags/bools to mask the date. There's nothing in exif or xmp that fully supports this; it would have to stay inside Immich. Probably something like an advance options drop down when you go to set a date and a warning that it will not work outside Immich. That way users can still decide how to handle date/time but it can be presented as unknown in Immich. I really like the way iso8601:2-2019 handled it; they have three flags: uncertain, approximate, unspecified. As long as this sound good as a general approach we can probably start working out finer details: timeline display, searching, data type in database, etc.

**Reply by dkbarn** [reply]
A date range would be more versatile than a date mask. You cannot express uncertainty such as "1930/40s" or "Summer 1967" or "school year 1988/9" or "when grandma was a teenager" with a date mask, but you can with upper and lower bound dates.

**Reply by Silver0006** [reply]
> A date range would be more versatile than a date mask. You cannot express uncertainty such as "1940s" or "Summer 1967" or "school year 1988/9" with a date mask, but you can with upper and lower bound dates. 

I agree with the point about date ranges. But you totally can express uncertainty with a date mask. If you provide 1940 as a date and you mark 1110 the 0 would represent uncertainty. It might be I'm just stuck on the ideas in iso8601:2-2019. I definitely agree we should do date ranges though. I guess my question is, how would you store the data and how would a user interact with it?

**Reply by jrasm91** [reply]
Probably just nullable `asset.estimatedRangeStart` and `asset.estimatedRangeEnd` columns on the asset table that are full datetime objects. The web/mobile app can probably show have a dropdown for fuziness like day, week, month, year, decade, custom, etc. and then you pick a specific value. For example, fuziness: Year, value: 2025 (taken sometime in 2025). Or, fuziness: custom start: 2025-06, end 2025-08 (summer 2025). The UI can be updated to help the user specify their range, but it always gets sent to the api/database as a full date (midnight of the first day of june and 11:59:59 on last day of august, etc.)

**Comment by Silver0006** [comment]
https://github.com/immich-app/immich/discussions/8266
